docs/thrift.md

```diff
@@ -0,0 +1,205 @@
+
+-   **Service** refers to individual `service`s defined in the Thrift IDL.
+-   **System** refers to the whole system being designed in the Thrift IDL. A
+    system may consist of multiple services.
+
+A Thrift IDL for a system may contain multiple Thrift services that partition
+the different concerns of the system. For example,
+
+```thrift
+service UserService {
+    UserId createUser(1: UserDetails details)
+    void verifyEmailAddress(1: UserId userId, 2: VerificationToken token)
+}
+
+service PostService {
+    PostId submitPost(1: UserId userId, 2: PostInfo post)
+}
+```
+
+There are two ways to consume such a multi-service system:
+
+-   Set up a separate server on a different port and/or a different machine for
+    each service in the system. Consumers specify the different hosts/ports
+    while constructing their clients.
+-   Multiplex the different services behind the same server.
+
+I'll focus on the second approach in this section because the first approach is
+not very different from having separate systems for each service.
+
+As mentioned in [`arg1`][], each service method will be registered with the
+TChannel server in the format `{serviceName}::{methodName}`. For the example
+above, we'll have 3 endpoints: `UserService::createUser`,
+`UserService::verifyEmailAddress`, and `PostService::submitPost`.
+
+Callers must use the full endpoint name when making requests. For example,
+
+```javascript
+send({
+    service: "UserService",   // < This is the TChannel service name
+    endpoint: "PostService::submitPost",
+    // ...
+})
+```
+
+For convenience, client implementations may allow omission of the
+`{serviceName}::` prefix for the common case where the TChannel service name
+matches the Thrift service name. For example,
+
+```javascript
+send({service: "UserService", endpoint: "createUser"})
+// The implementation should translate this to,
+send({service: "UserService", endpoint: "UserService::createUser"})
+```
+
+Service Inheritance
+-------------------
+
+Thrift supports the concept of service inheritance. For example,
+
+```thrift
+service BaseService {
+    bool isHealthy()
+}
+
+service UserService extends BaseService {
+    // ...
+}
+
+service PostService extends BaseService {
+    // ...
+}
+```
+
+In case of service inheritance, we don't want the "parent" service's methods to
+be registered under its name. In the example above, we *don't* want
+`BaseService::isHealthy` registered. Instead, `UserService::isHealthy` and
+`PostService::isHealthy` must be registered.
+
+To do this, the server code responsible for registering endpoints must first
+denormalize the Thrift file into a set of "leaf" services that contain all
+methods -- including inherited ones.
+
+Uncaught Exceptions
+-------------------
+
+For uncaught server-side exceptions that are not defined in the Thrift IDL,
+server implementations must respond with a TChannel `error` message with error
```

Comment:
This is not a `must`. Servers can crash in which case the response is an ECONNRESET or equivelant.

Also is it possible for unknown operational errors that are not defined in the thrift IDL to be create a CallResponse whose code is `1` or non-zero.